### PR TITLE
Add ArcUpgradeScheme

### PR DESCRIPTION
**Pull request by ben-kaufman**

### CODE REVIEW

File: contracts/schemes/ArcUpgradeScheme.sol [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+
+        emit ProposalExecutedByVotingMachine(address(avatar), _proposalId, _decision);
+        return true;
+    }
+
+    /**
+    * @dev execution of proposals after it has been decided by the voting machine
+    * @param _proposalId the ID of the voting in the voting machine
+    */
+    function execute(bytes32 _proposalId) public {
```
Comment (orenyodfat) [inline contracts/schemes/ArcUpgradeScheme.sol]: why this is public function ?
is there a chance that the genericCall will revert on the execution time and will not refer after it ?

File: contracts/schemes/ArcUpgradeScheme.sol [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+    function proposeUpgrade(
+        uint64[3] memory _packageVersion,
+        bytes32[] memory _contractsNames,
+        address[] memory _contractsToUpgrade,
+        string memory _descriptionHash)
+    public
+    returns(bytes32)
+    {
+        require(package.hasVersion(_packageVersion), "Specified version doesn't exist in the Package");
+        for (uint256 i = 0; i < _contractsToUpgrade.length; i++) {
```
Comment (orenyodfat) [inline contracts/schemes/ArcUpgradeScheme.sol]: I think we should get the max possible length which can be executed gas wised and cap the length of this array with that

File: contracts/schemes/ArcUpgradeScheme.sol [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+            proposalId,
+            _packageVersion,
+            _contractsNames,
+            _contractsToUpgrade,
+            _descriptionHash
+        );
+        return proposalId;
+    }
+
+    function bytes32ToStr(bytes32 x) private pure returns (string memory) {
```
Comment (orenyodfat) [inline contracts/schemes/ArcUpgradeScheme.sol]: this bit complex util function is also at https://github.com/daostack/arc/blob/arc-factory/contracts/utils/DAOFactory.sol#L234 .
how about make it in a lib and use it in both places ?

File: contracts/schemes/ArcUpgradeScheme.sol [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+import "@openzeppelin/upgrades/contracts/Initializable.sol";
+import "@openzeppelin/upgrades/contracts/application/Package.sol";
+import "@openzeppelin/upgrades/contracts/application/ImplementationProvider.sol";
+
+
+/**
+ * @title ArcUpgradeScheme.
+ * @dev  A scheme for proposing updating the versions DAOs contracts
+ */
+contract ArcUpgradeScheme is VotingMachineCallbacks, ProposalExecuteInterface, Initializable {
```
[the remark this answers is not in the document]
Comment (ben-kaufman) [reply contracts/schemes/ArcUpgradeScheme.sol]: We already have UpgradeScheme

File: contracts/schemes/ArcUpgradeScheme.sol (hunk 1 above)
Comment (ben-kaufman) [reply contracts/schemes/ArcUpgradeScheme.sol]: Only chance I think is if the Avatar is not the admin of the contract proposed. Not sure it’s necessary to have this public but I just sticked to the same way you made GenericScheme

File: contracts/schemes/ArcUpgradeScheme.sol (hunk 2 above)
Comment (ben-kaufman) [reply contracts/schemes/ArcUpgradeScheme.sol]: 20 seems to be around the max, so will cap it there ok?

File: contracts/schemes/ArcUpgradeScheme.sol [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+    event ProposalExecutedByVotingMachine(
+        address indexed _avatar,
+        bytes32 indexed _proposalId,
+        int256 _param
+    );
+
+    event ProposalDeleted(address indexed _avatar, bytes32 indexed _proposalId);
+
+    // Details of a voting proposal:
+    struct UpgradeProposal {
```
[the remark this answers is not in the document]
Comment (ben-kaufman) [reply contracts/schemes/ArcUpgradeScheme.sol]: fixed

File: contracts/schemes/ArcUpgradeScheme.sol [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+import "@daostack/infra-experimental/contracts/votingMachines/VotingMachineCallbacksInterface.sol";
+import "../votingMachines/VotingMachineCallbacks.sol";
+import "@openzeppelin/upgrades/contracts/Initializable.sol";
+import "@openzeppelin/upgrades/contracts/application/Package.sol";
+import "@openzeppelin/upgrades/contracts/application/ImplementationProvider.sol";
+
+
+/**
+ * @title ArcUpgradeScheme.
+ * @dev  A scheme for proposing updating the versions DAOs contracts
```
[the remark this answers is not in the document]
Comment (ben-kaufman) [reply contracts/schemes/ArcUpgradeScheme.sol]: fixed

File: contracts/schemes/ArcUpgradeScheme.sol (hunk 3 above)
Comment (ben-kaufman) [reply contracts/schemes/ArcUpgradeScheme.sol]: moved to a lib

File: contracts/schemes/ArcUpgradeScheme.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (ben-kaufman) [reply contracts/schemes/ArcUpgradeScheme.sol]: Fixed

File: contracts/schemes/ArcUpgradeScheme.sol (hunk 1 above)
Comment (orenyodfat) [reply contracts/schemes/ArcUpgradeScheme.sol]: it should not be public

File: test/upgradescheme.js [hunk 7]
```diff
@@ -1,239 +1,444 @@
...
+            web3.utils.fromAscii("Avatar"),web3.utils.fromAscii("DAOToken"),web3.utils.fromAscii("Reputation"),
+            web3.utils.fromAscii("Avatar"),web3.utils.fromAscii("DAOToken"),web3.utils.fromAscii("Reputation"),
+            web3.utils.fromAscii("Avatar"),web3.utils.fromAscii("DAOToken"),web3.utils.fromAscii("Reputation"),
+            web3.utils.fromAscii("Avatar"),web3.utils.fromAscii("DAOToken"),web3.utils.fromAscii("Reputation"),
+            web3.utils.fromAscii("Avatar"),web3.utils.fromAscii("DAOToken"),web3.utils.fromAscii("Reputation"),
+            web3.utils.fromAscii("Avatar"),web3.utils.fromAscii("DAOToken"),web3.utils.fromAscii("Reputation"),
+            web3.utils.fromAscii("Avatar"),web3.utils.fromAscii("DAOToken"),web3.utils.fromAscii("Reputation"),
+            web3.utils.fromAscii("Avatar"),web3.utils.fromAscii("DAOToken"),web3.utils.fromAscii("Reputation")
+          ],
+          [
```
Comment (orenyodfat) [inline test/upgradescheme.js]: how about filling this array in a for loop ?

File: test/upgradescheme.js [hunk 8]
```diff
@@ -1,239 +1,444 @@
...
+          web3.utils.fromAscii("Avatar"),web3.utils.fromAscii("DAOToken"),web3.utils.fromAscii("Reputation"),
+          web3.utils.fromAscii("Avatar"),web3.utils.fromAscii("DAOToken"),web3.utils.fromAscii("Reputation"),
+          web3.utils.fromAscii("Avatar"),web3.utils.fromAscii("DAOToken"),web3.utils.fromAscii("Reputation"),
+          web3.utils.fromAscii("Avatar"),web3.utils.fromAscii("DAOToken"),web3.utils.fromAscii("Reputation"),
+          web3.utils.fromAscii("Avatar"),web3.utils.fromAscii("DAOToken"),web3.utils.fromAscii("Reputation"),
+          web3.utils.fromAscii("Avatar"),web3.utils.fromAscii("DAOToken"),web3.utils.fromAscii("Reputation"),
+          web3.utils.fromAscii("Avatar"),web3.utils.fromAscii("DAOToken"),web3.utils.fromAscii("Reputation")
+        ],
+        [
+          testSetup.org.avatar.address, testSetup.org.token.address, testSetup.org.reputation.address,
```
Comment (orenyodfat) [inline test/upgradescheme.js]: for loop ?

File: test/upgradescheme.js (hunk 7 above)
Comment (ben-kaufman) [reply test/upgradescheme.js]: done

File: test/upgradescheme.js (hunk 8 above)
Comment (ben-kaufman) [reply test/upgradescheme.js]: done

File: contracts/schemes/ArcUpgradeScheme.sol (hunk 1 above)
Comment (ben-kaufman) [reply contracts/schemes/ArcUpgradeScheme.sol]: You mean here or also in the GenericScheme?

Comment (orenyodfat) [reply contracts/schemes/ArcUpgradeScheme.sol]: here. 
in GenericScheme it is different case where there is a chance that the target contract call will revert though the arc tx will not due to the nature of contract.call ...

The case in this scheme is different as there is no way the target contract can revert.

File: contracts/schemes/ArcUpgradeScheme.sol (hunk 2 above)
Comment (orenyodfat) [reply contracts/schemes/ArcUpgradeScheme.sol]: does the number 60 verified/tested ?

Comment (ben-kaufman) [reply contracts/schemes/ArcUpgradeScheme.sol]: It is tested that 60 works. I think more is possible theoretically but not much so just rounded it down.

File: contracts/schemes/ArcUpgradeScheme.sol (hunk 1 above)
Comment (ben-kaufman) [reply contracts/schemes/ArcUpgradeScheme.sol]: ok.

File: test/upgradescheme.js [hunk 9]
```diff
@@ -196,8 +196,8 @@ contract('UpgradeScheme', function(accounts) {
       let contractsNames = [];
       let contractsToUpgrade = [];
       for (let i = 0; i < 21; i++) {
```
Comment (orenyodfat) [inline test/upgradescheme.js]: how about make it 60 ? like in the contract limit

Comment (ben-kaufman) [reply test/upgradescheme.js]: the loop adds 3 at a time

File: contracts/schemes/UpgradeScheme.sol [hunk 10]
```diff
@@ -61,106 +73,106 @@ contract UpgradeScheme is Initializable, VotingMachineCallbacks, ProposalExecute
...
-                bytes4 permissions = controller.schemesPermissions(address(this));
-                require(
-                controller.registerScheme(proposal.upgradeContract, permissions)
+    function executeProposal(bytes32 _proposalId, int256 _decision)
+    external
+    onlyVotingMachine(_proposalId)
+    returns(bool) {
+        Proposal storage proposal = organizationProposals[_proposalId];
+        require(proposal.exist, "must be a live proposal");
+        require(proposal.passed == false, "cannot execute twice");
```
Comment (orenyodfat) [inline contracts/schemes/UpgradeScheme.sol]: do we now need these ?

File: contracts/schemes/UpgradeScheme.sol [hunk 11]
```diff
@@ -61,106 +73,106 @@ contract UpgradeScheme is Initializable, VotingMachineCallbacks, ProposalExecute
...
+                    abi.encodeWithSignature("upgradeTo(address)", updatedImp),
+                    0
                 );
-                if (proposal.upgradeContract != address(this)) {
-                    require(controller.unregisterSelf());
-                }
             }
+
+            delete organizationProposals[_proposalId];
+            emit ProposalDeleted(address(avatar), _proposalId);
```
Comment (orenyodfat) [inline contracts/schemes/UpgradeScheme.sol]: this can be taken out of the if statement ..

File: contracts/schemes/UpgradeScheme.sol [hunk 12]
```diff
@@ -61,106 +73,106 @@ contract UpgradeScheme is Initializable, VotingMachineCallbacks, ProposalExecute
...
+            emit ProposalExecuted(address(avatar), _proposalId);
+        } else {
+            delete organizationProposals[_proposalId];
+            emit ProposalDeleted(address(avatar), _proposalId);
```
Comment (orenyodfat) [inline contracts/schemes/UpgradeScheme.sol]: same as above

File: contracts/schemes/UpgradeScheme.sol [hunk 13]
```diff
@@ -61,106 +73,106 @@ contract UpgradeScheme is Initializable, VotingMachineCallbacks, ProposalExecute
...
                 );
-                if (proposal.upgradeContract != address(this)) {
-                    require(controller.unregisterSelf());
-                }
             }
+
+            delete organizationProposals[_proposalId];
+            emit ProposalDeleted(address(avatar), _proposalId);
+            emit ProposalExecuted(address(avatar), _proposalId);
+        } else {
```
Comment (orenyodfat) [inline contracts/schemes/UpgradeScheme.sol]: no need else

File: contracts/schemes/UpgradeScheme.sol (hunk 10 above)
Comment (ben-kaufman) [reply contracts/schemes/UpgradeScheme.sol]: Exists I think yes, but I removed the passed property

**Comment by dOrgJelli** [comment]
Amazing! :clap: :heart_eyes:

File: contracts/schemes/UpgradeScheme.sol [hunk 14]
```diff
@@ -61,106 +66,99 @@ contract UpgradeScheme is Initializable, VotingMachineCallbacks, ProposalExecute
...
-                controller.registerScheme(proposal.upgradeContract, permissions)
+    function executeProposal(bytes32 _proposalId, int256 _decision)
+    external
+    onlyVotingMachine(_proposalId)
+    returns(bool) {
+        Proposal storage proposal = organizationProposals[_proposalId];
+        require(proposal.exist, "must be a live proposal");
+
+        if (_decision == 1) {
+            proposal.exist = false;
```
Comment (orenyodfat) [inline contracts/schemes/UpgradeScheme.sol]: so this is another protection ..along with onlyVotingMachine .. let it be :)

**Pull request merged**